iree/base/math.h

```diff
@@ -117,99 +129,267 @@ IREE_FORCEINLINE int CountLeadingZeros64(uint64_t n) {
+static inline int iree_math_count_ones_u32(uint32_t n) {
+  n -= ((n >> 1) & 0x55555555u);
+  n = ((n >> 2) & 0x33333333u) + (n & 0x33333333u);
+  return (int)((((n + (n >> 4)) & 0x0F0F0F0Fu) * 0x01010101u) >> 24);
 }
 
 // Returns the number of 1 bits in a 64 bit value.
-IREE_FORCEINLINE int CountOnes64(uint64_t n) {
-  return CountOnes32(n >> 32) + CountOnes32(n & 0xffffffff);
+static inline int iree_math_count_ones_u64(uint64_t n) {
+  return iree_math_count_ones_u32(n >> 32) +
+         iree_math_count_ones_u32(n & 0xFFFFFFFFu);
 }
 
+//==============================================================================
+// Rounding and alignment
+//==============================================================================
+// There are certain platforms - mostly those with poorer quality compilers or
+// more restricted instruction sets - where we want to avoid the clz path as
+// it is emulated and instead we use some bit-twiddling hacks. On other
+// platforms it's the opposite - they may emulate clz but doing so saves
+// dozens of bytes that otherwise would have been the shift/or tree.
+//
+// Which to choose is entirely determined by fiddling on godbolt for the
+// target platform: https://godbolt.org/z/h4vPzo
+
 // Rounds up the value to the nearest power of 2 (if not already a power of 2).
-IREE_FORCEINLINE int RoundUpToNearestPow2(int n) {
-  return n ? ~0u >> LeadingZeros(n) : 1;
+// For 32-bit numbers this only supports values <= 2^31; higher will wrap.
+static inline uint32_t iree_math_round_up_to_pow2_u32(uint32_t n) {
+#if 0    // golf required; can be bloated
+  const uint32_t i = (n != 1);
+  return (1 + i) << ((iree_math_count_leading_zeros_u32(n - i) ^ 31));
+#elif 0  // golf required; can be bloated
+  return n == 1 ? 1u : 2u << ((iree_math_count_leading_zeros_u32(n - 1) ^ 31));
+#else
+  // https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+  n--;
+  n |= n >> 1;
+  n |= n >> 2;
+  n |= n >> 4;
+  n |= n >> 8;
+  n |= n >> 16;
+  return n + 1;
+#endif  // 1
+}
+
+// Rounds up the value to the nearest power of 2 (if not already a power of 2).
+// For 64-bit numbers this only supports values <= 2^63; higher will wrap.
+static inline uint64_t iree_math_round_up_to_pow2_u64(uint64_t n) {
+#if 0    // golf required; can be bloated
+  const uint64_t i = (n != 1);
+  return (1 + i) << ((iree_math_count_leading_zeros_u64(n - i) ^ 63));
+#elif 0  // golf required; can be bloated
+  return n == 1 ? 1ull
+                : 2ull << ((iree_math_count_leading_zeros_u64(n - 1) ^ 63));
+#else
+  // https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+  n--;
+  n |= n >> 1;
+  n |= n >> 2;
+  n |= n >> 4;
+  n |= n >> 8;
+  n |= n >> 16;
+  n |= n >> 32;
+  return n + 1;
+#endif  // 1
+}
+
+//==============================================================================
+// Pseudo-random number generators (PRNGs): **NOT CRYPTOGRAPHICALLY SECURE*
```

Comment:
this is a bit of an understatement. more descriptive would be "very cheap PRNGs, high performance, low quality".

The c++ `<random>` header defines some useful PRNGs. If we're building our own here, that means we know something specific to our intended applications that requires different PRNGs. What is it?

(The default PRNG to beat is linear congruential, since that's a single instruction, SIMDs trivially, is clearly uniformly distributed over the whole integer type, and is readily available as `std::minstd_rand`).

Reply:
This is for C (so no <random>) and a very specific set of use cases where code-size and cost/iteration matters; the goal is to have something that gives an adequate distribution with commensurate cost: in this case, 16 bytes of state and 5-6 instructions per generation. This particular PRNG was designed for the use cases we are applying it to and is used in quite a few schedulers (including the java, mesa, solaris, and dtrace thread schedulers). It's also used quite a bit more broadly in things like numba and numpy. You don't have the follow on PR for context but it's literally used to select a number between 0-63 to bias a work-stealing queue - and even then this may be _too good_ for that use case ;)

Reply:
So, for such low-quality 6-bit valus, a 8bit LCG gives you a ratio of 1 instruction (intrinsic = vmulq_u8) for 16 generated random bytes on ARM NEON.  Can you beat that? :-)

Reply:
oo neat, didn't know that existed! that's probably worth using for that particular worker selection case!
(after I land threading, I'm not looking to golf any more until it's in ;)